action/protocol/poll/governance_protocol.go

```diff
@@ -303,3 +381,91 @@ func (p *governanceChainCommitteeProtocol) getGravityHeight(ctx context.Context,
 	)
 	return p.electionCommittee.HeightByTime(blkTime)
 }
+
+func (p *governanceChainCommitteeProtocol) getKickoutBlackList(epochNum uint64) (map[string]bool, error) {
+	kickoutList, err := p.kickoutListByEpoch(p.sr, epochNum)
+	if err != nil {
+		return nil, err
+	}
+	kickoutMapping := make(map[string]bool, 0)
+	for _, addr := range kickoutList {
+		kickoutMapping[addr] = true
+	}
+	return kickoutMapping, nil
+}
+
+func (p *governanceChainCommitteeProtocol) calculateKickoutBlackList(
+	ctx context.Context,
+	epochNum uint64,
+) ([]string, error) {
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	blkCtx := protocol.MustGetBlockCtx(ctx)
+	rp := rolldpos.MustGetProtocol(bcCtx.Registry)
+	englishEpochNum := rp.GetEpochNum(config.English)
+
+	var unqualifiedDelegates []string
+	if epochNum <= englishEpochNum+p.kickoutEpochPeriod {
+		// if epoch number is smaller than EnglishHeightEpoch+K(kickout period), calculate it one-by-one (initialize).
+		round := epochNum - englishEpochNum // 0, 1, 2, 3 .. K
+		for {
+			if round == 0 {
+				break
+			}
+			// N-1, N-2, ..., N-K
+			numBlks, produce, err := p.productivityByEpoch(ctx, epochNum-round)
+			if err != nil {
+				return nil, err
+			}
+			if round == 1 {
+				// The current block is not included, so that we need to add it to the stats
+				numBlks++
+				produce[blkCtx.Producer.String()]++
+			}
+			expectedNumBlks := numBlks / uint64(len(produce))
+			for addr, actualNumBlks := range produce {
+				if actualNumBlks*100/expectedNumBlks < p.productivityThreshold {
+					unqualifiedDelegates = append(unqualifiedDelegates, addr)
+				}
+			}
+			round--
+		}
+		return unqualifiedDelegates, nil
+	}
+
+	// Blacklist[N] = Blacklist[N-1] - BlackList[N-K] + Low-productivity-list[N-1]
+	prevBlacklist, err := p.kickoutListByEpoch(p.sr, epochNum-1)
+	if err != nil {
+		return nil, err
+	}
+	prevKthBlacklist, err := p.kickoutListByEpoch(p.sr, epochNum-p.kickoutEpochPeriod)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Blacklist[N] = Blacklist[N - 1] - low_productivity_list[N-K] + low_productivity_list[N - 1]
where Blacklist[N] is a map[string]int (int indicates the number of low_productivity_list[N - k] == true, where k is in [0, K-1])

Reply:
K = 2
   low_p_list.  Blacklist
1: [A B C] -> [A:1 B:1 C:1]
2: [B D] -> [A:1 B:2 C:1 D:1]
3: [E F] -> [B D E F] vs [A:0 B:1 C:0 D:1 E:1 F:1]

Reply:
BlackList[N] = counter under-performance delegates between [N-K, N-1] 
So, BlackList[N] = BlackList[N-1] - low_productivity_list[N-K-1] + low_productivity_list[N-1]